display: stm32_ltdc: fix double fb, implement rotation & suspend behavior

emillindq opened the pull request:
The current implementation of a double frame buffer is flawed,
especially considering that LVGL does multiple small writes, essentially
making double buffering pointless as only part of the frame is rendered
when the buffer is swapped. One workaround has been to use a VDB_SIZE of
100% (e.g. STM32H7B3I_DK) and as such have LVGL perform the double
buffering. However, for instance software rotation doesn't work in this
configuration and the entire frame needs to be refreshed on every write.

This commit checks if the write is partial. If this is the case, just
write the partial buffer and return. If the write is final, the buffers
are swapped and the new buffer is copied to the old one.

In addition, an intermediate buffer is used on copying if the frame
buffer is located in SDRAM, to speed up copying of the frame.

danieldegrasse commented:
It looks like these changes are all related to performance optimization- in that case, why not use a VDB size of 100? If the LTDC peripheral is being used with a video mode display (which as I understand it, is how it is designed to be used) then the full frame will be refreshed every VSYNC interval anyways. At least on the NXP systems with a video engine like the LTDC, a VDB size of 100 significantly improved performance, as no copy was required (the frame buffer pointer can just be swapped)

emillindq commented:
[danieldegrasse](https://github.com/danieldegrasse)
I suppose you are implying that your LVGL VDB is located in your on-board ram? With high-resolution displays, this doesn't always fit and requires an SD-RAM to which access is much much slower compared to internal RAM, implying a lot of small small writes from LVGL. IIRC this also implies setting the `LV_Z_FULL_REFRESH` symbol, with which LVGL software rotation doesn't work.

They do it the way you describe in the stm32h7b3i_dk, which works fine, but that board doesn't require display rotation.

The way we have it setup is a 10% LVGL VDB in on-board RAM, which is flushed all at once to the SD-RAM

danieldegrasse commented:
> I suppose you are implying that your LVGL VDB is located in your on-board ram? With high-resolution displays, this doesn't always fit and requires an SD-RAM to which access is much much slower compared to internal RAM, implying a lot of small small writes from LVGL. IIRC this also implies setting the LV_Z_FULL_REFRESH symbol, with which LVGL software rotation doesn't work.

At least on the systems I've used (iMX.RT), the framebuffer is in external RAM, yeah- but the performance gained by avoiding the memory copies between framebuffers vastly outweighs any performance hit from using SDRAM.

I wasn't aware that setting `LV_FULL_REFRESH` would disable display rotation- on the iMX.RT it was possible to enable rotation in combination with the PXP engine (I believe STM32 has a similar peripheral, the DMA2D). That being said, if this is needed to get your application to work I'm not opposed to it- I was just suggesting potential methods to improve performance. Will defer to @erwango or anyone else from ST on if this change makes sense overall though

erwango commented:
> @erwango or anyone else from ST on if this change makes sense overall though

Here is the feedback from a colleague more knowledgeable than myself on this topic, even if not fully expert. It doesn't answer fully the question but hopefully gives food for thoughts:

I looked at the pull request, especially the first commit. My feeling about this point is already that it seems to address two distinct issues that should therefore be handled in two different commits: 
- the first concerns the ability to handle incomplete writes from the caller of the display framework. That is, not doing a buffer swap (and a duplication of the two buffers) with each write but only when the application indicates that this is NOT an incomplete write. This is done by keeping the return 0 if the frame is incomplete, and by moving the memcpy to the end of the function, only AFTER the buffer swap has taken place. 
- the second, and I think this is more the subject of the discussion here, is the use of this intermediate copy buffer. I admit I am not familiar with these points regarding memory access latency when using SDRAM. If it greatly improves performance, why not, I would say, I don't find it very elegant, but well. However, having had memory space issues on the L4, I am a bit sensitive to creating arrays like this and I wonder if it should not be optional as well. It's an optimization, in a way speed vs size, so I would find it good to want to do without this intermediate buffer. We can also ask ourselves the question of the potential use of a DMA in such a case, right? If we consider that the display can also be quite large, it quickly becomes a few hundred kilobytes for a framebuffer, so maybe the penalty of starting a DMA is compensated. I don't know.

I have no experience with LVGL for the moment, I saw that it was a module used to do the composition above the framebuffers, but I have never used it yet. I understand that the VDBs mentioned are somewhat buffer memory areas that this composer will use to prepare before sending everything to the framebuffers. That is why the person commenting suggests having a VDB of 100 so that there is never an incomplete write in the framebuffer.

But that does not change the fact that I think it is still necessary to support incomplete writes (aka this first patch I mentioned above), because incomplete writes are not something limited only to LVGL but rather a functionality of the display framework itself. The sample application uses it as well.

emillindq commented:
@danieldegrasse DMA2D or STM32's LTDC peripheral doesn't support hardware-accelerated rotation. Regardless, in the latest version of LVGL they have moved software rotation to a function call which is called from **LVGL's display driver**, instead of applying it globally. So the cleanest way now to support rotation is probably to in fact implement sw rotation in  this driver.

Which in turn would make a VDB size of 100% viable...

@erwango the intermediate buffer gives slightly better performance but I need to revisit this entire PR from the start given the latest changes in LVGL mentioned above (am currently porting our code to v4.1.0-rc2) and your comments. I'll get back to you.

emillindq committed:
display: stm32_ltdc: fix double fb, implement rotation & suspend behavior

This commit implements and fixes a number of things:
- If a double frame buffer is used, the frames are swapped only when
  the entire frame has been rendered
- The new frame buffer is copied to the old frame buffer to support
  partial updates of the frame buffer
- Rotation of 90° is implemented as rotate-on-write to frame buffer
- If the device is suspended, the frame buffers are not swapped since
  the ISR is not running and the sem would never be given, causing
  a lock
- Miscellaneous optimizations, especially if an SDRAM is used to store
  the frame buffer

Signed-off-by: Emil Lindqvist <emil@lindq.gr>

emillindq commented:
@erwango @danieldegrasse please take a look on the newly pushed changes. Commit message:
```
This commit implements and fixes a number of things:
- If a double frame buffer is used, the frames are swapped only when
  the entire frame has been rendered
- The new frame buffer is copied to the old frame buffer to support
  partial updates of the frame buffer
- Rotation of 90° is implemented as rotate-on-write to frame buffer
- If the device is suspended, the frame buffers are not swapped since
  the ISR is not running and the sem would never be given, causing
  a lock
- Miscellaneous optimizations, especially if an SDRAM is used to store
  the frame buffer
```
I have verified it is working with an stm32h7b3i_dk, please see https://github.com/emillindq/zephyr/tree/stm32h7b3i_dk_test

![image](https://github.com/user-attachments/assets/cbb7aabd-8e14-4380-9c20-bc48c5b1052b) ![image](https://github.com/user-attachments/assets/f150e481-87ec-4eaf-9052-f7a49619ec21)

danieldegrasse commented:
If there is no hardware support for rotation than I'm ok with this approach- it looks like you have done some performance optimization to enable this to work reasonably well. I'm going to leave review to @erwango though, because I'm not familiar with this hardware

markussmST commented on drivers/display/display_stm32_ltdc.c:
```diff
@@ -171,58 +175,151 @@ static int stm32_ltdc_write(const struct device *dev, const uint16_t x,
-				dst = data->frame_buffer + data->frame_buffer_len;
-			}
+	dst = data->frame_buffer;
+	old = data->frame_buffer + data->frame_buffer_len;
```
What is the use of this? It will create an entire new buffer, but only if running single buffer, since both `CONFIG_STM32_LTDC_FB_NUM == 2` and `CONFIG_STM32_LTDC_FB_NUM == 0 `are handled separately.
The memory is not allocated, so it will result in an unsafe write.
Fortunately, the write should never happen because `data->front_buf == pend_buf` on line 282 will usually be true in the case of single buffering, unless the user wants to be able to change between two separate buffers for some reason.

emillindq replied:
Now I only assign old if it's used, and check for `old == NULL` before copying

markussmST commented on drivers/display/display_stm32_ltdc.c:
```diff
@@ -171,58 +175,151 @@ static int stm32_ltdc_write(const struct device *dev, const uint16_t x,
 			return -ENOTSUP;
 		}
+		/* Use buf as ltdc frame buffer directly if it is identical to it. */
+		data->front_buf = buf;
```
This should be `data->pend_buf` instead, then the swap in the isr will work, which I think is important for users who want to manage double buffering outside the driver. It may also solve the problem you observe with e.g. the H7B3.

emillindq replied:
That works, thanks!

kartben commented on drivers/display/display_stm32_ltdc.c:
```diff
@@ -171,58 +175,151 @@ static int stm32_ltdc_write(const struct device *dev, const uint16_t x,
+	 *   - memcpy: 7fps
+	 *   - this: 11fps
+	 */
+	static uint8_t tmp[CONFIG_STM32_LTDC_INTERMEDIATE_BUF_SIZE];
```
not sure if there can be multiple LTDC instances but this static buffer would be an issue as this makes stm32_ltdc_write non-reentrant

markussmST commented:
Have you considered keeping track of the areas that were updated since the last swap and just copying the relevant areas to the old framebuffer after the swap instead of the entire buffer? I think that would bring a large performance gain for most frames.

emillindq commented:
Apologies for delays, finally got some time to look into this again.

> Have you considered keeping track of the areas that were updated since the last swap and just copying the relevant areas to the old framebuffer after the swap instead of the entire buffer? I think that would bring a large performance gain for most frames.

For sure, but I can't imagine the complexity this would bring. For example, LVGL might render two or more completely different parts of the screen not adjacent to each other. Suggestions welcome though!

markussmST commented:
> Apologies for delays, finally got some time to look into this again.
> 
> > Have you considered keeping track of the areas that were updated since the last swap and just copying the relevant areas to the old framebuffer after the swap instead of the entire buffer? I think that would bring a large performance gain for most frames.
> 
> For sure, but I can't imagine the complexity this would bring. For example, LVGL might render two or more completely different parts of the screen not adjacent to each other. Suggestions welcome though!

I'm not sure it would be that much more complexity, just store the coordinates of each update in an array and iterate through it at the end, copying the pixels bounded by the coordinates. You would of course have to keep track of the offsets, so it would become a memcpy of the block width per image line in the block, but I bet it would be much faster in almost all situations.

emillindq commented:
>I'm not sure it would be that much more complexity, just store the coordinates of each update in an array and iterate through it at the end, copying the pixels bounded by the coordinates. You would of course have to keep track of the offsets, so it would become a memcpy of the block width per image line in the block, but I bet it would be much faster in almost all situations.

Maybe a different PR in that case? This is big enough as it is, and it would require either an array of an unbounded size, alt. dynamic memory allocation (I imagine)
